### DISCUSSION TITLE
Spherical Phase Space file

### DISCUSSION THREAD

**Discussion by mojtaba1365** [nrc-cnrc/EGSnrc#984] [Questions] [answered]
Hi,
I am dealing with spherical geometry and I wonder if it is possible to generate a spherical phsp file around a sphere surrounded by another sphere. 

I used Method 2 to define a scorer to generate a phsp file containing all particles exiting the smaller sphere and entering the larger one, but it seems the scorer took into account all particles inside the smaller sphere!

Any comments?

**Comment by rtownson** [comment]
The egs++ phase-space scoring code is fairly new, so this could be a bug. If your inner sphere is region 1, and the outer sphere is region 2, then this should work (but I haven't tried) to score the particles on the surface between the two spheres. How are you testing the phase-space to confirm?

```
:start ausgab object definition:
      :start ausgab object:
          library                  = egs_phsp_scoring
          name                     = some_name
          output format            = IAEA
          particle type            = all
          score multiple crossers  = no
          output directory         = /something/
          from regions             = 1
          to regions               = 2
      :stop ausgab object:
:stop ausgab object definition:
```

Duplicate post:
https://www.reddit.com/r/EGSnrc/comments/12ir8qi/spherical_phase_space_file/

**Reply by mojtaba1365** [reply]
Thanks for responding,

In my case, the inner region is 0 (R=5 cm) and the outer one is 1 (R=10 cm) and I defined the scoring plane similar to your suggestion (in EGS format not IAEA).

I attached a screenshot of the geometry and tracks in the regions.

I ran BEAMDP_GUI to analyze the phsp file using X_Y scatter pattern of particles using annular distribution and limited the analysis region to 4.99 to 10 cm having only the particles exiting the inner sphere. But as a result, I observed just 6 dots in the graph while as you can see in the screenshot there are much more particles exiting the inner region! I am not sure but it may consider the annular region only at Z=0 and not a 3D scanning of the phsp file.

It may not be a good idea to assess this phsp info by X_Y scatter plots due to the spherical geometries. I want to be sure it is possible to define a spherical scoring plane in this case and that plane is thoroughly scoring only particles exiting the region.
I would appreciate it if you point out any tools or methods to verify the phsp content.

![ki0s6gy61hta1](https://user-images.githubusercontent.com/30004572/231508493-d153966b-a265-4a19-9f1c-db5a12c39fa6.jpg)

**Reply by rtownson** [reply]
First, you will want to switch to IAEA format output, otherwise all Z information is lost. The X-Y scatter plot will only be looking at r^2=x^2+y^2, so along the z=0 axis. BEAMDP doesn't really have good 3D analysis tools. For a sanity check, you could use its function to list particle parameters for the first 20 particles, and then do some hand calculations to confirm that the x/y/z positions all correspond to the correct scoring radius.

**Reply by mojtaba1365** [reply]
As you mentioned, using EGSnrc format the Z coordinate would be missing and I see only the X and Y coordinates in the parameter list, so I can not manually confirm manually the particle positions scored in the phsp file. Moreover, the BEAMDP does not support the IAEA format to extract the parameter list. 
Any other solution?

**Reply by rtownson** [reply, marked as answer]
Yes, BEAMDP can read IAEA phase-space files. It just doesn't print the z position. Here are some changes to beamdp to fix that. Make these changes and recompile beamdp.

```diff
diff --git a/HEN_HOUSE/omega/progs/beamdp/beamdp.mortran b/HEN_HOUSE/omega/progs/beamdp/beamdp.mortran
index 39f2fac0..a3dca742 100644
--- a/HEN_HOUSE/omega/progs/beamdp/beamdp.mortran
+++ b/HEN_HOUSE/omega/progs/beamdp/beamdp.mortran
@@ -1270,13 +1270,13 @@ IF(MSTART=0 & CHOICE1=11)[
    IF(MODE_RW='MODE0')[
       WRITE(6,:LIST-FORMAT0:);
       :LIST-FORMAT0: FORMAT(/' ENERGY  IQ     X',
-      5X, '  Y',4X,'  U',4X,'  V',4X,'  W',4X,'WEIGHT',6X,
+      5X, '  Y',4X,'  Z',4X,'  U',4X,'  V',4X,'  W',4X,'WEIGHT',6X,
       ' LATCH (set=1, not set=0)'/);
    ]
    ELSEIF(MODE_RW='MODE2')[
       WRITE(6,:LIST-FORMAT01:);
       :LIST-FORMAT01: FORMAT(/' ENERGY  IQ     X',
-      5X, '  Y',4X,'  U',4X,'  V',4X,'  W',4X,' ZLAST   WEIGHT',6X,
+      5X, '  Y',4X,'  Z',4X,'  U',4X,'  V',4X,'  W',4X,' ZLAST   WEIGHT',6X,
       ' LATCH (set=1, not set=0)'/);
    ]
    CALL READ_DATA;"to read the ph-sp file"
@@ -2140,14 +2140,14 @@ integer lnblnk1;
           IF(IIII>=NUMBERST)[
           IF(MODE_RW='MODE0')[
              WRITE(6,:LIST-FORMAT1:)
-             EI,IQ,XIN,YIN,UIN,VIN,WIN,WEIGHT,(LATCHJ(I),I=1,29);
+             EI,IQ,XIN,YIN,ZIN,UIN,VIN,WIN,WEIGHT,(LATCHJ(I),I=1,29);
-             :LIST-FORMAT1:FORMAT(F7.3,I4,2F8.3,3F7.3,1PE10.3,
+             :LIST-FORMAT1:FORMAT(F7.3,I4,3F8.3,3F7.3,1PE10.3,
              2X,$LATCH_NUMBER_OF_BITSI1,1X,24I1);
              ]
           ELSEIF(MODE_RW='MODE2')[
              WRITE(6,:LIST-FORMAT11:)
-             EI,IQ,XIN,YIN,UIN,VIN,WIN,ZLAST,WEIGHT,(LATCHJ(I),I=1,29);
+             EI,IQ,XIN,YIN,ZIN,UIN,VIN,WIN,ZLAST,WEIGHT,(LATCHJ(I),I=1,29);
-             :LIST-FORMAT11:FORMAT(F7.3,I4,2F8.3,3F7.3,F8.3,1PE10.3,
+             :LIST-FORMAT11:FORMAT(F7.3,I4,3F8.3,3F7.3,F8.3,1PE10.3,
              2X,$LATCH_NUMBER_OF_BITSI1,1X,24I1);
              ]
           ]
@@ -2180,10 +2180,10 @@ integer lnblnk1;
        ]
 
       :LIST-FORMAT02: FORMAT(/' ENERGY  IQ     X',
-      5X, '  Y',4X,'  U',4X,'  V',4X,'  W',4X,'WEIGHT',6X,
+      5X, '  Y',4X,'  Z',4X,'  U',4X,'  V',4X,'  W',4X,'WEIGHT',6X,
       ' LATCH (set=1, not set=0)'/);
       :LIST-FORMAT03: FORMAT(/' ENERGY  IQ     X',
-      5X, '  Y',4X,'  U',4X,'  V',4X,'  W',4X,' ZLAST   WEIGHT',6X,
+      5X, '  Y',4X,'  Z',4X,'  U',4X,'  V',4X,'  W',4X,' ZLAST   WEIGHT',6X,
       ' LATCH (set=1, not set=0)'/);
```

**Reply by mojtaba1365** [reply]
Thanks for your valuable comment!
I could extract the Z coordinates from IAEA format and after manual calculation of all Rs, it looks all particles scored in the plane are at R=5 cm. It is amazing!

**Comment by mojtaba1365** [comment]
Hi @rtownson,

As you recommended, I generated a couple of phase space files in the IAEA format. I analyzed them by BEAMdp and it seems the particles' data were scored well in the files.
However, I tried to use the spherical phsp files as sources to obtain the dose distribution beyond the central sphere, but this error popped up:

In EGS_Application::initSimulation()
EGS_PhspSource::openFile: the file m_phsp.1.IAEAphsp is not a MODE0 or MODE2 file
EGS_PhspSource: errors while opening the phase space file m_phsp.1.IAEAphsp
createSource(phsp source): the input is not sufficient to create a valid source
EGS_ObjectFactory::createObjects(): 1 errors occured while creating objects

It seems there is an issue with the reading of the IAEAphsp files since I tried for EGSnrc format and everything worked fine...

Any comment?

**Reply by rtownson** [reply]
For IAEA phase-spaces you need to use this different class (`library = iaea_phsp_source`):

https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/classIAEA__PhspSource.html

**Reply by mojtaba1365** [reply]
Thanks a lot.
